### PR TITLE
Private harbor fix

### PR DESCRIPTION
**Pull request by at88mph**
Added ability to specify `x-skaha-registry-auth` to request header in format `base64Encode(username:secret)`.
Cleanup for Checkstyle
Added to Desktop and passed to Desktop App

### CODE REVIEW

File: deployment/helm/skaha/desktop-template/start-software-sh.template [hunk 1]
```diff
@@ -123,7 +125,7 @@ get_status() {
     if [ -z "${TOKEN}" ]; then
         handle_error "[skaha] No credentials to call back to Skaha with."
     else
-        curl_out=`curl -s -L -k --header "x-auth-token-skaha: ${TOKEN}" https://(HOST)/skaha/(SKAHA_API_VERSION)/session/${VNC_PW}/app/$1`
+        curl_out=`curl -s -L -k "${CURL_APP_SESSION_HEADERS}" https://(HOST)/skaha/(SKAHA_API_VERSION)/session/${VNC_PW}/app/$1`
```
Comment (brianmajor) [inline deployment/helm/skaha/desktop-template/start-software-sh.template]: I think the registry credentials are only needed for the POST calls, but I suppose it doesn't hurt to have them in here too.

File: deployment/helm/skaha/desktop-template/start-software-sh.template [hunk 2]
```diff
@@ -89,7 +91,7 @@ prompt_user() {
             if [ -z "${TOKEN}" ]; then
               handle_error "[skaha] No credentials to call back to Skaha with."
             else
-              app_id=`curl -s -L -k --header "x-auth-token-skaha: ${TOKEN}" -d "image=(IMAGE_ID)" --data-urlencode "param=(NAME)" https://(HOST)/skaha/(SKAHA_API_VERSION)/session/${VNC_PW}/app`
+              app_id=`curl -s -L -k "${CURL_APP_SESSION_HEADERS}" -d "image=(IMAGE_ID)" --data-urlencode "param=(NAME)" https://(HOST)/skaha/(SKAHA_API_VERSION)/session/${VNC_PW}/app`
```
Comment (brianmajor) [inline deployment/helm/skaha/desktop-template/start-software-sh.template]: Don't both credentials need to be sent in?  (TOKEN and REGISTRY_AUTH)?

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Constructor to use the Base64 encoded value to obtain the credentials for an Image Registry.
+     *
+     * @param encodedValue The Base64 encoded String.  Never null.
+     * @param host  The registry host for these credentials.
+     * @return ImageRegistryAuth instance.  Never null.
+     */
+    public static ImageRegistryAuth fromEncoded(final String encodedValue, final String host) {
```
Comment (brianmajor) [inline skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java:39]: static?

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public String getUsername() {
+        return username;
+    }
+
+    public byte[] getSecret() {
+        return secret;
+    }
+
+    public boolean isSecretValid() {
+        return !new String(this.secret).startsWith("\ufffd");
```
Comment (brianmajor) [inline skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: Is this character sequence related to the encoding?  Can the encoding be checked instead?

File: skaha/src/main/java/org/opencadc/skaha/session/PostAction.java [hunk 5]
```diff
@@ -482,39 +489,35 @@ private String validateImage(String imageID, String type) throws Exception {
...
-                if (type == null) {
-                    return image.getTypes().iterator().next();
-                } else {
-                    if (image.getTypes().contains(type)) {
-                        return type;
-                    } else {
-                        throw new IllegalArgumentException("image/type mismatch: " + imageID + "/" + type);
-                    }
-                }
+        final String imageRegistryHost = getRegistryHost(imageID);
```
Comment (brianmajor) [inline skaha/src/main/java/org/opencadc/skaha/session/PostAction.java:486]: The 'harborHosts' config was acting as a list of accepted/valid registries.  I think we need to preserve that concept?

Comment (brianmajor) [reply skaha/src/main/java/org/opencadc/skaha/session/PostAction.java:486]: Or maybe the logic was just shifted to elsewhere?

File: skaha/src/main/java/org/opencadc/skaha/session/PostAction.java [hunk 6]
```diff
@@ -482,39 +489,35 @@ private String validateImage(String imageID, String type) throws Exception {
...
+        log.debug("Image is located at " + imageRegistryHost);
+
+        final Image image = getPublicImage(imageID);
+        final String validatedType;
+
+        // Private images are also missing from this list.
+        if (image == null) {
+            log.debug("Image " + imageID + " missing from cache...");
+            // Absence of type is assumed to be headless
```
Comment (brianmajor) [inline skaha/src/main/java/org/opencadc/skaha/session/PostAction.java]: The comment is right but I don't see how that is supported in the code (type==null)

File: skaha/src/main/java/org/opencadc/skaha/utils/CommandExecutioner.java [hunk 7]
```diff
@@ -102,6 +82,54 @@ public static void execute(final String[] command, final OutputStream standardOu
...
+        // harbor invalidates secrets with the Unicode replacement characters 'fffd'.
+        if (registryAuth.isSecretValid()) {
+            // create new secret
+            final String[] createCmd = new String[] {
+                "kubectl", "--namespace", K8SUtil.getWorkloadNamespace(), "create", "secret", "docker-registry",
+                secretName,
+                "--docker-server=" + registryAuth.getHost(),
+                "--docker-username=" + registryAuth.getUsername(),
+                "--docker-password=" + new String(registryAuth.getSecret())
+            };
```
Comment (brianmajor) [inline skaha/src/main/java/org/opencadc/skaha/utils/CommandExecutioner.java]: This is nice

**Review by brianmajor — CHANGES_REQUESTED** [review]

File: deployment/helm/skaha/desktop-template/start-software-sh.template (hunk 2 above)
Comment (at88mph) [reply deployment/helm/skaha/desktop-template/start-software-sh.template]: Yes.  The `CURL_APP_SESSION_HEADERS` variable contains both:
https://github.com/opencadc/science-platform/pull/711/files/3a06b2af919698c588b162eecda1b8833fdc3361#diff-2de56bc8187f4afad440423d15d2324cccf69f03706a79b18809168df3053697R7

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 4 above)
Comment (at88mph) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: This is leftover from some old check.  This is just preserving some old check that was in place, but I'm not sure what its origins are.

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 3 above)
Comment (at88mph) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java:39]: Yes, in a builder/singleton pattern style since the encoded value won't be preserved, so putting into a constructor didn't seem appropriate:
```java
final ImageRegistryAuth registryAuth = ImageRegistryAuth.fromEncoded(encoded, "docker.io");
```

File: skaha/src/main/java/org/opencadc/skaha/session/PostAction.java (hunk 5 above)
Comment (at88mph) [reply skaha/src/main/java/org/opencadc/skaha/session/PostAction.java:486]: That's still true.  This is to isolate what host the provided Image was for.  It gets parsed from the ID.

File: skaha/src/main/java/org/opencadc/skaha/session/PostAction.java (hunk 6 above)
Comment (at88mph) [reply skaha/src/main/java/org/opencadc/skaha/session/PostAction.java]: Ah, true, will fix.  The type is defaulted here:
https://github.com/opencadc/science-platform/blob/3a06b2af919698c588b162eecda1b8833fdc3361/skaha/src/main/java/org/opencadc/skaha/session/PostAction.java#L207

File: deployment/helm/skaha/desktop-template/start-software-sh.template (hunk 1 above)
Comment (at88mph) [reply deployment/helm/skaha/desktop-template/start-software-sh.template]: That's true and I got carried away with the copy & pasting.  I think in the interest of isolating the use of that authorization key, I'll make sure it's only used where needed.

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 4 above)
Comment (brianmajor) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: Oh yeah.  When CLISecrets would expire in Harbor, shaka would see them starting with these characters, which basically means 'not encodable'.  This check won't be needed anymore because the users are providing them directly (copying them from harbor) so won't get these weird values that we were seeing through the Harbor API.

**Comment by brianmajor** [comment]
@shinybrar Could you please take a look at this PR as well?

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+     * @param encodedValue The Base64 encoded String.  Never null.
+     * @param host  The registry host for these credentials.
+     * @return ImageRegistryAuth instance.  Never null.
+     */
+    public static ImageRegistryAuth fromEncoded(final String encodedValue, final String host) {
+        if (!StringUtil.hasText(encodedValue)) {
+            throw new IllegalArgumentException("Encoded auth username and key is required.");
+        }
+
+        final String decodedValue = new String(Base64.decode(encodedValue));
```
Comment (shinybrar) [inline skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: Should we explicitly use `StandardCharsets.UTF_8` when converting between byte arrays and strings?

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    public String getHost() {
+        return host;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public byte[] getSecret() {
+        return secret;
```
Comment (shinybrar) [inline skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java:69]: Is the returned secret mutable or immutable. Do we have a practice to always keep as immutable values to deter against parsing anomalies?

**Review by shinybrar — COMMENTED** [review]
Keep in mind,
- I have no Java Background
- These are big classes of code to get lost in

Comments
- We could use `StandardCharsets.UTF_8` explicitly when converting between byte arrays and strings for encoding and decoding.
- There is one edgecase which still goes unchecked when the user provides a valid registry server, valid registry credentials, but malformed image/tag names for private images. This will lead to a 200 for the user, but the job will fail to launch. After discussions with @at88mph -- this is a known edgecase, which is currently bound by the limitation of our system.
- Since we perform kubectl bash command underneath the hood, the sting stream created in `CommandExecutioner` can create an escape scenario, if the client secret has a escape character in the secret. 
- We should add unittests for the new code added

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 9 above)
Comment (shinybrar) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java:69]: @at88mph confirmed, this secret is always immutable as described by the `final` prefix in Java.

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 4 above)
Comment (at88mph) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: OK, removed.

File: skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java (hunk 8 above)
Comment (shinybrar) [reply skaha/src/main/java/org/opencadc/skaha/registry/ImageRegistryAuth.java]: Here is a possible implementation,
```java
final String decodedValue = new String(Base64.decode(encodedValue), StandardCharsets.UTF_8);
```

**Review by shinybrar — DISMISSED** [review]
LGTM

**Comment by brianmajor** [comment]
@at88mph @shinybrar - Does the removal of the "headless" type affect the behaviour of the API at all?
